### ISSUE TITLE
LLaMA-Adapter: Efficient Fine-tuning of Language Models with Zero-init Attention

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/paper-reviews#211]
# TL; DR

- Introduce parameter and compute efficient adaptation technique for LLaMa that empowers it by tuning to follow instructions.

# Motivation

- Instruction fine-tuning greatly elevates LLM's practical performance, but can be costly and compute-intensive.
- Stanford Alpaca demonstrated promising performances but requires 7B full-scale finetuning.

# Formulation

![image](https://user-images.githubusercontent.com/46308407/230596957-de76e715-8c1a-4275-a6ac-1749ef0a9705.png)

### Learnable prompts

- Introduce K learnable prefix tokens to incorporate instruction following capacity for top L layers. (K=10, L=30, total 32 Tr layers in practice)
- Start by applying zero-gating for these prefix tokens to minimize disturbance to the pretrained knowledge.
![image](https://user-images.githubusercontent.com/46308407/230597166-dccfbba5-153a-4821-a6ff-cbc890612180.png)
- Learnable gating parameter is initialized as zero.
- Two separate softmax operation ensures that the pretrained representation is not affected in the beginning.

### Multi-modal Extension

- LLaMa-Adapter presents a simple way to further incorporate multi-modality to conventional LLM like LLaMa.
- Form context embedding of an image using a vision model like CLIP-ViT, and project it to have same dimension with prompt tokens, and apply element-wise addition.

![image](https://user-images.githubusercontent.com/46308407/230597403-7aa82f65-1c91-4424-b40c-5d786f70d8f8.png)
![image](https://user-images.githubusercontent.com/46308407/230597389-eaf929f7-10e1-4ce5-8631-67083ac86d09.png)

# Evaluation

![image](https://user-images.githubusercontent.com/46308407/230597447-014119d8-a698-45cc-900a-d0c98c9782f9.png)
![image](https://user-images.githubusercontent.com/46308407/230597466-c3342e65-7a10-4c7d-ad59-657e88454840.png)
![image](https://user-images.githubusercontent.com/46308407/230597482-d45940ce-f992-4a36-8391-c2bdfcaf11f1.png)
![image](https://user-images.githubusercontent.com/46308407/230597507-8a15c7e4-3f5c-46cb-8477-c50b1af5072a.png)

# Inspiration

- Pretrained foundation models are everywhere. How to effectively use them is the key.
